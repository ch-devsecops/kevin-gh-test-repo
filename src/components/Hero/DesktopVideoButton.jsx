import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Button, useThemeContext, Media } from '@honda-canada/design-system-react';

const VideoButtonIcon = ({ isHovering }) => {
  const colors = useThemeContext('colors');
  const color = isHovering ? colors.black : colors.white;

  return (
    <svg
      style={{ marginLeft: '8px', position: 'relative', top: '1px' }}
      width="8"
      height="12"
      viewBox="0 0 8 12"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M0 0V12L8 6L0 0Z" fill={color} />
    </svg>
  );
};

VideoButtonIcon.propTypes = {
  isHovering: PropTypes.bool,
};

const DesktopVideoButton = ({ video, gtmTags }) => {
  const [videoHover, setVideoHover] = useState(false);

  if (!video) return null;

  return (
    <Media greaterThanOrEqual="smallDesktop">
      <Button
        styling="secondaryDark"
        onClick={video.onPlay}
        onMouseEnter={() => setVideoHover(true)}
        onMouseLeave={() => setVideoHover(false)}
        aria-label={video.ariaLabel}
        mt={[2, 4]}
        {...gtmTags}
      >
        {video.buttonLabel}
        <VideoButtonIcon isHovering={videoHover} />
      </Button>
    </Media>
  );
};

DesktopVideoButton.propTypes = {
  video: PropTypes.shape({
    ariaLabel: PropTypes.string,
    onPlay: PropTypes.func,
    buttonLabel: PropTypes.string,
  }),
};

export default DesktopVideoButton;
