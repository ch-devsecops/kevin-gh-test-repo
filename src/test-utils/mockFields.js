/* eslint-disable max-len */
/* eslint-disable max-classes-per-file */

// FIXME: jsdoc popup not working on test files
/** Class representing a mock data. */
class MockData {
  /**
   * Add a field.
   * @param {string} key - key of the field obj: i.e. 'title'
   * @param {Object} value - value of the field obj: i.e. { value: 'some text' }.
   */
  add(key, value) {
    this[key] = value;
  }
}

/**
 * Class representing mock fields object.
 * @extends MockData
 *
 * available method: add()
 *
 * default fields: [anchorId, title, bodyText, image, ctaLink, ctaType, ctaIcon, gtmCategory, gtmTitle, contentTitle, subtitle]
 *
 * note: we can add more commonly used fields in the class declaration if needed
 */
class MockFields extends MockData {
  constructor() {
    super();

    this.title = { value: 'Title' };
    this.bodyText = { value: 'BodyText' };
    this.image = {
      value: { src: 'mock.png', alt: 'mock alt' },
    };
    this.ctaLink = {
      value: {
        href: 'www.source.ca',
        title: 'cta aria label',
        text: 'CTA Text',
      },
    };
    this.ctaType = {
      value: 'Tertiary',
    };
    this.ctaIcon = {
      fields: {
        value: {
          value: 'arrowRight',
        },
      },
    };
    this.gtmCategory = {
      fields: {
        value: {
          value: 'Others',
        },
      },
    };
    this.gtmTitle = {
      value: 'gtm title',
    };
    this.contentTitle = {
      value: 'Content Title',
    };
    this.subtitle = {
      value: 'Content Subtitle',
    };
    this.seoH1 = {
      value: '',
    };
    this.anchorId = {
      value: 'anchor-id',
    };
    this.ctaLabel = {
      value: 'CTA Label',
    };
  }
}

/**
 * Class representing a mock rendering object
 * @extends MockData
 *
 * available method: add()
 */
class MockRendering extends MockData {
  /**
   * Create a mock rendering object
   * @param {string} componentName - adds componentName
   */
  constructor(componentName) {
    super();

    this.componentName = componentName;
  }
}

export { MockFields, MockRendering };
