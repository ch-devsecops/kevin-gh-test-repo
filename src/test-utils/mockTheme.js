/**
 * Mock styled-components theme, if you want to test how your component
 * implements theme styles. Import as-is, or spread in any
 * additional values you need.
 *
 * Example usage:

  it('should apply the bgColor param', () => {
    const { colors } = MockTheme;
    const { container } = render(
      <MyComponent
        fields={mockFields}
        params={{ bgColor: 'Gray' }}
      />,
    );
    const wrapper = container.querySelector('section');

    expect(wrapper).toHaveStyleRule('background-color', colors.grey[5]);
  });
 */

import { HONDA_THEME_NAME } from '../utils/constants';

const mockTheme = {
  letterSpacings: {
    sm: '0.5px',
    md: '1px',
  },
  lineHeights: {
    sm: '22px',
    md: '26px',
  },
  colors: {
    white: '#fff',
    transparent: 'transparent',
    grey: ['#707070', '#C7C7C7', '#D5D5D5', '#EAEAEA', '#F8F8F8', '#F5F5F5'],
    black: '#2C2C2C',
    red: '#DA2C2E',
    redShades: ['rgba(200, 16, 46, 0.1)'],
    aodaRed: '#D92629',
    lightRed: '#EC9597',
    darkRed: '#A21C1D',
    blue: '#334C96',
    darkBlue: '#002D62',
    lightBlue: '#00447826',
    yellow: ['#F9CA52', '#EDAB00', '#8D763F'],
    fadedBlue: '#ADB7D5',
    primary: '#DA2C2E',
    primaryAoda: '#D92629',
    primaryDark: '#A21C1D',
    primaryLight: '#EC9597',
    secondary: '#2C2C2C',
    secondaryLight: '#C7C7C7',
    tertiary: '#002D62',
    tertiaryLight: '#334C96',
    tertiaryDisabled: '#ADB7D5',
    danger: '#DA2C2E',
    warning: '#2C2C2C',
    warningLight: '#EAEAEA',
    success: '#334C96',
    successLight: '#d9e0e7',
    disabled: '#C7C7C7',
    typographyDefault: '#2C2C2C',
  },
  space: [0, 4, 8, 12, 16, 24, 32, 48, 64, 96, 128, 144, 192, 256],
  borders: [0, '1px solid', '2px solid', '3px solid'],
  breakpoints: ['768px', '992px', '1200px'],
  fontWeights: {
    regular: 'normal',
    bold: 700,
  },
  fontSizes: {
    sm: '12px',
  },
  shadows: {
    smooth: {
      md: '0.1px 0.3px 0.3px rgba(0, 0, 0, 0.02),0.3px 0.7px 0.9px rgba(0, 0, 0, 0.028),0.7px 1.4px 1.8px rgba(0, 0, 0, 0.035),1.5px 2.9px 3.7px rgba(0, 0, 0, 0.047),4px 8px 10px rgba(0, 0, 0, 0.07)',
    },
    box: {
      subNav: '0px 2px 4px rgba(0, 0, 0, 0.1)',
    },
  },
  zIndices: {
    accordionItemTitle: 200,
    modal: 13001,
    sectionNav: 13000,
  },
  transitionDuration: {
    t1: '1s',
  },
  transitionTimingFunction: {
    default: 'ease-in-out',
    out: 'ease-out',
    in: 'ease-in',
  },
  name: HONDA_THEME_NAME,
  fonts: {
    default: 'Helvetica',
    bold: 'Cooper Black',
    heading: 'Roboto',
  },
  header: {
    desktop: {},
    mobile: {},
  },
  icons: {},
  logIconWarnings: false,
};

export default mockTheme;
