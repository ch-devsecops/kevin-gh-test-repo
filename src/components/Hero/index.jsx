import React from 'react';
import { Button, Copy, Hero as DesignSystemHero, Optional, useMediaQueries } from '@honda-canada/design-system-react';
import { Link as SitecoreLink } from '@sitecore-jss/sitecore-jss-react';
import isValidUrlOrRelativeUrl from '@honda-canada/js-utilities/lib/isValidUrlOrRelativeUrl';
import HeroBadge from '../HeroBadge';
import { wrapJSSFields } from '../../utils/wrapJSSFields';
import { getVideoProps, mapGTMCategory } from '../../utils/sitecoreFields';
import { getTitleComponent, stripMarkdownHeading, styledCompiler } from '../../utils/markdown';
import DesktopVideoButton from './DesktopVideoButton';
import { getGtmTagValue } from '../../utils/gtmEvents';
import { useAppName } from '../../utils/sitecoreContext';
import { PSP_SITE_NAME } from '../../utils/constants';

export const getHorizontalAlignment = (contentPosition = '') => {
  if (contentPosition?.includes('Left')) return 'left';
  if (contentPosition?.includes('Right')) return 'right';
  if (contentPosition?.includes('Center')) return 'center';

  return 'left';
};

export const getVerticalAlignment = (contentPosition = '') => {
  if (contentPosition?.includes('Top')) return 'top';
  if (contentPosition?.includes('Bottom')) return 'bottom';

  return 'bottom';
};

const Hero = ({ fields, params, rendering }) => {
  const appName = useAppName();
  const { isMobile } = useMediaQueries();
  if (!fields) return null;

  const {
    seoH1,
    title,
    subtitle,
    desktopImage,
    mobileImage,
    ctaUrl,
    ctaType,
    ariaLabel,
    badgeImage,
    gtmTitle,
    gtmCategory,
    gtmInteractionType,
    gtmModelName,
    gtmTrimName,
    gtmBodyStyle,
    videoUrl,
  } = wrapJSSFields(fields);

  const video = getVideoProps(
    videoUrl?.value?.href,
    videoUrl?.value?.title,
    'close video modal',
    videoUrl?.value?.text,
  );

  const horizontalAlignment = getHorizontalAlignment(params?.contentPosition);
  const verticalAlignment = getVerticalAlignment(params?.contentPosition);
  const isShort = params?.heroStyleType === 'Short';
  const gtmTags = {
    category: mapGTMCategory(gtmCategory),
    type: rendering?.componentName,
    title: gtmTitle?.value,
    interactionType: gtmInteractionType?.value || 'cta: click',
    modelName: gtmModelName?.value,
    trimName: gtmTrimName?.value,
    bodyStyle: gtmBodyStyle?.value,
  };

  let ctaGtmTags = {
    'data-gtm-model': getGtmTagValue(gtmModelName?.value),
    'data-gtm-trim': getGtmTagValue(gtmTrimName?.value),
    'data-gtm-body-style': getGtmTagValue(gtmBodyStyle?.value),
    'data-gtm-interaction-type': getGtmTagValue(gtmInteractionType?.value),
    'data-gtm-title': getGtmTagValue(gtmTitle?.value),
    'data-gtm-component-type': getGtmTagValue(rendering?.componentName),
  };

  switch (appName) {
    case PSP_SITE_NAME:
      ctaGtmTags = {
        'data-gtm-component-type': rendering?.componentName,
        'data-gtm-interaction-type': getGtmTagValue(gtmInteractionType?.value),
        'data-gtm-title': gtmTitle?.value,
      };
      break;
    default:
      break;
  }

  const TitleComponent = getTitleComponent(title?.value);

  const buttons = () => (
    <>
      {ctaUrl?.getProp('url') && (
        <Button
          as={SitecoreLink}
          aria-label={ctaType?.value?.title}
          field={ctaUrl?.field}
          styling={`${ctaType.value?.toLowerCase() || 'secondary'}Dark`}
          mt={[2, 4]}
          mr={[0, video ? 3 : 0]}
          {...ctaGtmTags}
        />
      )}
      <DesktopVideoButton video={video} gtmTags={ctaGtmTags} />
    </>
  );

  const imageAltText = (isMobile ? mobileImage?.getProp('alt') : desktopImage?.getProp('alt')) || ariaLabel?.value;
  const desktopImageSource = isValidUrlOrRelativeUrl(desktopImage?.getProp('src')) ? desktopImage?.getProp('src') : '';
  const mobileImageSource = isValidUrlOrRelativeUrl(mobileImage?.getProp('src')) ? mobileImage?.getProp('src') : '';

  return (
    <DesignSystemHero
      mobileImageSource={mobileImageSource}
      desktopImageSource={desktopImageSource}
      horizontalAlignment={horizontalAlignment}
      verticalAlignment={verticalAlignment}
      imageAltText={imageAltText}
      globalAriaLabels={ariaLabel?.value}
      isShort={isShort}
      video={video}
      gtmTags={gtmTags}
      buttons={buttons}
    >
      <HeroBadge image={badgeImage} horizontalAlignment={horizontalAlignment} />
      <Optional when={title?.field?.value}>
        <TitleComponent as={seoH1?.value === 'Title' ? 'h1' : 'p'} color="white" mt={['8px', '16px']} width="100%">
          {styledCompiler(stripMarkdownHeading(title?.value))}
        </TitleComponent>
      </Optional>

      <Copy
        as={seoH1?.value === 'Subtitle' ? 'h1' : 'p'}
        color="white"
        mt="2"
        mr={horizontalAlignment === 'right' ? 0 : 'auto'}
        ml={horizontalAlignment === 'left' ? 0 : 'auto'}
        maxWidth={['none', '400px']}
        lineHeight={['22px', '26px']}
      >
        {styledCompiler(stripMarkdownHeading(subtitle?.value))}
      </Copy>
    </DesignSystemHero>
  );
};

export default Hero;
