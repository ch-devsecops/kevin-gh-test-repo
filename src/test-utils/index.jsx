import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { SitecoreContext } from '@sitecore-jss/sitecore-jss-react';
import { BrowserRouter } from 'react-router-dom';
import { OptionalVideoModalContextProvider } from '@honda-canada/design-system-react';
import UserLocationProvider from '@honda-canada/user-location';
import { configureAxe } from 'jest-axe';
import ModelExplorationContextProvider from '../components/ModelExplorationContext';
import ModelFiltersContextProvider from '../components/ModelFiltersContext';
import { MockFields, MockRendering } from './mockFields';
import MockTheme from './mockTheme';
import 'intersection-observer';
import { ACURA_THEME_NAME } from '../utils/constants';

const generateJssWrapper = context =>
  // eslint-disable-next-line react/display-name
  function ({ children }) {
    return (
      <SitecoreContext
        context={{
          language: 'en',
          site: { name: ACURA_THEME_NAME },
          ...context?.sitecoreContext,
        }}
        componentFactory={{}}
      >
        <BrowserRouter>
          <UserLocationProvider>
            <ThemeProvider theme={context?.theme || MockTheme}>
              <OptionalVideoModalContextProvider>
                <ModelExplorationContextProvider>
                  <ModelFiltersContextProvider>{children}</ModelFiltersContextProvider>
                </ModelExplorationContextProvider>
              </OptionalVideoModalContextProvider>
            </ThemeProvider>
          </UserLocationProvider>
        </BrowserRouter>
      </SitecoreContext>
    );
  };
/**
 * A custom implementation of the testing-library `render` function that
 * wraps tests in an typical JSS app's context providers.
 * Note that you can include a `context` object to set or override values
 * for SitecoreContext and ThemeContext.
 * @param {ReactElement} component - the component you want testing-library to render
 * @param {object} [options] - see https://testing-library.com/docs/react-testing-library/api/#render-options
 * @param {string} [options.context] - default values for SitecoreContext, ThemeProvider
 * @param {string} [options.context.sitecoreContext] - value for SitecoreContext
 * @param {string} [options.context.theme] - value for ThemeProvider
 * @example
 * const { getByText } = jssRender(
 *  <MyComponent />, { context: { sitecoreContext: { language: 'fr' } } }
 * );
 * @returns {function(): void} custom testing-library render function
 */
const jssRender = (component, options = {}) => {
  const { context } = options;
  // eslint-disable-next-line no-param-reassign
  delete options.context;

  return render(component, {
    wrapper: generateJssWrapper(context),
    ...options,
  });
};

const generateWrapper = context =>
  // eslint-disable-next-line react/display-name
  function ({ children }) {
    return (
      <BrowserRouter>
        <ThemeProvider theme={context?.theme || MockTheme}>
          <OptionalVideoModalContextProvider>{children}</OptionalVideoModalContextProvider>
        </ThemeProvider>
      </BrowserRouter>
    );
  };

/**
 * A custom implementation of the testing-library `render` function that
 * wraps tests in typical design-system context providers.
 * Note that you can include a `theme` object to set or override values
 * for ThemeContext.
 * @param {ReactElement} component - the component you want testing-library to render
 * @param {object} [options] - see https://testing-library.com/docs/react-testing-library/api/#render-options
 * @param {string} [options.theme] - default value for ThemeProvider
 * @example
 * const { getByText } = render(
 *  <MyComponent />, { theme: { space: { wrapperGap: '15px' } } }
 * );
 * @returns {function(): void} custom testing-library render function
 */
const customRender = (component, options = {}) => {
  const { theme } = options;
  // eslint-disable-next-line no-param-reassign
  delete options.theme;

  return render(component, {
    wrapper: generateWrapper(theme),
    ...options,
  });
};

// disable accessibility checks for errors that are usually trigged by content mistakes
const axe = configureAxe({
  rules: {
    'duplicate-id': { enabled: false },
    'heading-order': { enabled: false },
  },
});

export * from 'jest-axe';
export { axe };
export * from '@testing-library/react';
export { customRender as render, jssRender };
export { MockFields, MockRendering, MockTheme };
