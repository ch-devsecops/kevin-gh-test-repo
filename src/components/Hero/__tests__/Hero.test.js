import React from 'react';
import { axe, cleanup, jssRender } from 'test-utils';
import Hero from '..';

afterEach(cleanup);

const mockData = require('../../../../sample-mock-data/routes/hero/en.json');
const emptyMockData = require('../../../../sample-mock-data/routes/hero/empty_en.json');

describe('<Hero>', () => {
  test('should render correctly', () => {
    const { container } = jssRender(<Hero {...mockData} />);

    expect(container).toMatchSnapshot();
    expect(container.children[0]).toBeTruthy();
  });

  test('should return null if there is no fields', () => {
    const { container } = jssRender(<Hero fields={null} />);

    expect(container).toMatchSnapshot();
  });

  test('should return null if the data is empty', () => {
    const { container } = jssRender(<Hero {...emptyMockData} />);
    expect(container).toMatchSnapshot();
  });

  test('should have no accessibility violations', async () => {
    const { container } = jssRender(<Hero {...mockData} />);

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  });
});
