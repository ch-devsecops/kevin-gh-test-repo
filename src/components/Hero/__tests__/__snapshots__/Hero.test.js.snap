// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hero> should have no accessibility violations 1`] = `
<div>
  <div
    class="Box-sc-1bsf784-0 sc-1k44dfv-0 hyDNUx"
    data-gtm-category="contact us"
  >
    <span
      alt="Side shot of a Honda Accord parked on the road by trees during the fall season"
      aria-label="Side shot of a Honda Accord parked on the road by trees during the fall season"
      class="sc-1k44dfv-0 bgpmjE"
      role="img"
    />
    <div
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 gqxPaN"
    />
    <div
      class="Box-sc-1bsf784-0 Wrapper-n67dtg-0 sOsmN"
      height="100%"
    >
      <div
        class="Box-sc-1bsf784-0 ekugtK sc-1k44dfv-0 ivigvr"
        display="flex"
        height="100%"
        width="100%"
      >
        <div
          class="Box-sc-1bsf784-0 Column-t2p57o-0 sc-1k44dfv-0 koIMNU"
          width="1,0.75,0.5833333333333334"
        >
          <h1
            class="sc-1k44dfv-0 gwiNLS"
            color="white"
            width="100%"
          >
            <span>
              Contact Us
            </span>
          </h1>
          <p
            class="sc-1k44dfv-0 gOegea"
            color="white"
            font-family="default"
            font-weight="regular"
          >
            <span>
              Performance is in our DNA, even when it comes to customer service. If you have any questions or concerns relating to your Honda experience, you can reach us by one of the following methods.
            </span>
          </p>
          <div
            class="sc-1k44dfv-0 ldZtbH"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Hero> should render correctly 1`] = `
<div>
  <div
    class="Box-sc-1bsf784-0 sc-1k44dfv-0 hyDNUx"
    data-gtm-category="contact us"
  >
    <span
      alt="Side shot of a Honda Accord parked on the road by trees during the fall season"
      aria-label="Side shot of a Honda Accord parked on the road by trees during the fall season"
      class="sc-1k44dfv-0 bgpmjE"
      role="img"
    />
    <div
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 gqxPaN"
    />
    <div
      class="Box-sc-1bsf784-0 Wrapper-n67dtg-0 sOsmN"
      height="100%"
    >
      <div
        class="Box-sc-1bsf784-0 ekugtK sc-1k44dfv-0 ivigvr"
        display="flex"
        height="100%"
        width="100%"
      >
        <div
          class="Box-sc-1bsf784-0 Column-t2p57o-0 sc-1k44dfv-0 koIMNU"
          width="1,0.75,0.5833333333333334"
        >
          <h1
            class="sc-1k44dfv-0 gwiNLS"
            color="white"
            width="100%"
          >
            <span>
              Contact Us
            </span>
          </h1>
          <p
            class="sc-1k44dfv-0 gOegea"
            color="white"
            font-family="default"
            font-weight="regular"
          >
            <span>
              Performance is in our DNA, even when it comes to customer service. If you have any questions or concerns relating to your Honda experience, you can reach us by one of the following methods.
            </span>
          </p>
          <div
            class="sc-1k44dfv-0 ldZtbH"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Hero> should return null if the data is empty 1`] = `
<div>
  <div
    class="Box-sc-1bsf784-0 sc-1k44dfv-0 fOWcBp"
    data-gtm-category="Others"
  >
    <span
      class="sc-1k44dfv-0 hhMFLI"
      role="img"
    />
    <div
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 gqxPaN"
    />
    <div
      class="Box-sc-1bsf784-0 Wrapper-n67dtg-0 sOsmN"
      height="100%"
    >
      <div
        class="Box-sc-1bsf784-0 ekugtK sc-1k44dfv-0 ivigvr"
        display="flex"
        height="100%"
        width="100%"
      >
        <div
          class="Box-sc-1bsf784-0 Column-t2p57o-0 sc-1k44dfv-0 koIMNU"
          width="1,0.75,0.5833333333333334"
        >
          <p
            class="sc-1k44dfv-0 gOegea"
            color="white"
            font-family="default"
            font-weight="regular"
          >
            <span />
          </p>
          <div
            class="sc-1k44dfv-0 ldZtbH"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Hero> should return null if there is no fields 1`] = `<div />`;
